TxnCoordSender's heartbeat uses Context cancellation when available

danhhz opened the pull request:
Previously, the heartbeat was purely based on timeouts. Now, when given a context.Context with a non-nil Done channel (which indicates that it is cancellable), use that instead.

 Plumb a cancellable Context from a pgwire connection down to this heartbeat.

Closes #3031.

<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5447)

<!-- Reviewable:end -->

paperstreet commented:
@tamird @tschottdorf Before getting into the nitty-gritty review, I've got a few high level questions:

- To get tests working it seemed like I had to start passing the context into db.Txn (they were all ending up with leaked goroutines), which really exploded the size of this commit. Is there something else I could have done instead?
- After removing the `hasClientAbandonedCoord` check in the heartbeat loop, we're now relying entirely on the context lifetime for timing out the txn. How scary is this? I had to stick contexts in a lot of places to get this to work and a lot of them are still `context.TODO()`s. Do I have to fix them all before landing this?

tschottdorf commented:
I think `db.Txn(func)` can do without the closure - after all, you control the retryable, so you know when the action is over. Pseudocode (the real code looks somewhat different, but you get the idea):

```go
func (db *DB) Txn(closure func(*txn) error) error {
  ctx, cancel := context.WithCancel(context.Background())
  defer cancel()
  txn := NewTxn(ctx)
  for someCondition {
    err := closure(txn)
    if err .... {
    }
  }
}
```

Not sure if useful, but you could easily add `(*Stopper).Context(ctx) context.Context` which essentially creates a `context.WithCancel(ctx)` which cancels when the stopper is stopped - just in case you find yourself having a stopper in many situations which is supposed to also tear down the heartbeats (may not be the case).

tschottdorf commented:
What's not entirely clear is which lifetime a context passed to a transaction should have. Clearly, since you want it to control the heartbeat, the context (at least from client to TxnCoordSender) should be scoped to the lifetime of the transaction.

The v3conn has a context which is scoped to the lifetime of the connection. Executor needs to make sure we fork from that base contexts which are scoped to the lifetime of each individual transaction. I don't see that happening - I think you're mostly just passing a context around but never scoping it to a lifetime.

It's unfortunately hard to reason about the code since the raw plumbing is everywhere. Do you think you could pull this change apart? 1) plumb from `client.Txn` down, but don't change the heartbeat stuff yet (so this contains the bulk of your blind plumbing, but not much substance) 2) plumb from v3conn through executor to `*Txn`, making sure the lifetimes are taken care of. 3) replace the current heartbeat code.
(Or any variation of the above that makes sense to you).

tschottdorf commented:
Oh, and this needs a rebase.

bdarnell commented:
We need to keep the `hasClientAbandonedTxn` check as long as there are any transactions that don't go through pgwire (i.e. the admin CLI). It would be bad if pressing ctrl-c during an admin transaction left the coordinator heartbeating the transaction endlessly.

tschottdorf commented:
Agreed. I think the following would be a fairly legit use of `context.Context` in that setting:

```go
// util (or anywhere better):
type checkAbandonKey struct{}
func SetCheckAbandon(ctx) context.Context {
  return context.WithValue(ctx, checkAbandonKey{}, struct{}{})
}

func GetCheckAbandon(ctx) bool {
  _, ok := ctx.Value(checkAbandonKey{})
  return ok
}


// Executor:
ctx = util.SetCheckAbandon(ctx)
// Use ctx for all txns

// TxnCoordSender:
func heartbeatLoop(ctx context.Context, ...) {
  if util.GetCheckAbandon(ctx) { 
    // Use the traditional check
  }
}
```

The naming is bad and probably it should default to using the traditional method by default (and we always do the opposite in sql).

The bigger question remains: What are the semantics if the context is to be used for cancellation? After all, each individual request in the transaction passes in a `context.Context` and just using the one of the request which happens to start the heartbeat loop - that doesn't seem right.

danhhz committed:
TxnCoordSender's heartbeat uses Context cancellation when available

Previously, the heartbeat was purely based on timeouts. Now, when given a
context.Context with a non-nil Done channel (which indicates that it is
cancellable), use that instead.

Plumb a cancellable Context from a pgwire connection down to this heartbeat.

Fixes #3031.

paperstreet commented:
RFAL @bdarnell @tamird @tschottdorf

I'd like to add a test in txn_coord_sender_test for this. Any advice on where to start?

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

You should be able to crib from one of the txn abandonment tests. Perhaps look at the git blame for the "active abandon" functionality  - hopefully its tests are in the same commit.

---

Reviewed 31 of 31 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 669 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDdIjkXlBphq92acuTR:-KDdIjkXlBphq92acuTS:-1585284247) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b99c63de8bd0daaa85d5d4d88ccfc19a590994/kv/txn_coord_sender.go#L669)):*
this is the only thing that needs to happen under lock - can you write it as:
```
tc.Lock()
txn := tc.txns[txnID].txn.Clone()
tc.Unlock()
```

---

*[kv/txn_coord_sender.go, line 675 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDdIxT3FiLw4HivvUcf:-KDdIxT3FiLw4HivvUcg:233987579) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b99c63de8bd0daaa85d5d4d88ccfc19a590994/kv/txn_coord_sender.go#L675)):*
nit: you can write this as proceed := ...., or you can avoid the local variable entirely and use an expression in the `if` below.

---

*[sql/executor.go, line 29 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDdJJXtyDY9BC0_CFBj:-KDdJJXuENpX9xOEdhyR:1583801742) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b99c63de8bd0daaa85d5d4d88ccfc19a590994/sql/executor.go#L29)):*
nit: kill this line

---

*[sql/pgwire/v3.go, line 174 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDdJMTddkuIg1x8sG-I:-KDdJMTeJRE_syHm83YL:-1401450249) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b99c63de8bd0daaa85d5d4d88ccfc19a590994/sql/pgwire/v3.go#L174)):*
super nit: we usually call this method `cancel`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-:-KDdJ_TTnVYODjaY5d6o:1018486967)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
Done.

---

Review status: 3 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 669 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDdIjkXlBphq92acuTR:-KDdMeNiv2ns9IZsEbLh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b99c63de8bd0daaa85d5d4d88ccfc19a590994/kv/txn_coord_sender.go#L669)):*
Done.

---

*[kv/txn_coord_sender.go, line 675 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDdIxT3FiLw4HivvUcf:-KDdMenqo8SiLG0mA2sG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b99c63de8bd0daaa85d5d4d88ccfc19a590994/kv/txn_coord_sender.go#L675)):*
Done.

---

*[sql/executor.go, line 29 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDdJJXtyDY9BC0_CFBj:-KDdMl7tQRv1oj3hFX8j:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b99c63de8bd0daaa85d5d4d88ccfc19a590994/sql/executor.go#L29)):*
Done.

---

*[sql/pgwire/v3.go, line 174 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDdJMTddkuIg1x8sG-I:-KDdMuZGShTI-Bp4Pckb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/22b99c63de8bd0daaa85d5d4d88ccfc19a590994/sql/pgwire/v3.go#L174)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-:-KDeFiZGEDyELXTOYSB4:-499146009)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/txn_coord_sender_test.go, line 356 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeLAWireRrgkBpBaEL:-KDeLAWjD6K3OsudEihn:-1450006325) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/kv/txn_coord_sender_test.go#L356)):*
for the same reason that we didn't use `util.SucceedsSoon` in `TestTxnCoordSenderGCWithCancel`, you cannot use it here. This can instantly succeed, even when the reason no cleanup occurred is simply that not enough time had passed.

---

*[kv/txn_coord_sender_test.go, line 1270 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeLVL4SskJk70HiUtn:-KDeLVL51klvk75yujlt:-1911224481) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/kv/txn_coord_sender_test.go#L1270)):*
save future-self a dereference and inline this

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-:-KDeLdlmYN11DfIhOinH:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but the semantics about which context is used for the heartbeat loop need to be cleared up (that could mean documentation). I do think that there's a merit to doing this carefully - for example, the context is passed for each request, so a timeout for each request is reasonable but not doable because the current code would eagerly kill the transaction whenever the timeout fires. In the absence of that, passing the same one throughout the transaction will do the right thing, but is a little sketchy. For now I'm ok if you just document the behavior, though I'd prefer something better eventually.

---

Reviewed 3 of 31 files at r2, 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 412 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeMUkQoQe84giKsrhF:-KDeMUkQoQe84giKsrhG:-215995015) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/client/db.go#L412)):*
should make a real context outside with a cancellation, and defer it. The Txn doesn't outlive this method invocation.

---

*[kv/txn_coord_sender_test.go, line 571 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeNSSMO1V1k3sh5cTy:-KDeNSSMO1V1k3sh5cTz:789598808) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/kv/txn_coord_sender_test.go#L571)):*
Try running a manual invocation of the heartbeat loop, after which you know it would have been removed hat the check misfired. Same above where @tamird is commenting, I think. Otherwise, gotta make this timeout WAY shorter and deal with false successes when the heartbeat loop is slow.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-:-KDeOnh4VdmVBQmzFYwP:322221965)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 412 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeMUkQoQe84giKsrhF:-KDeQ4kiJe1mD8umc6ud:-699767888) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/client/db.go#L412)):*
Good call. That context should still be constructed from a context.TODO() though.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[client/db.go, line 412 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeMUkQoQe84giKsrhF:-KDeUEwmKVjPakSMN4HX:62476964) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/client/db.go#L412)):*
Why? The only lifetime of a context that makes sense with the current behavior is one for the lifetime of the transaction, and that's what we have here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447)*
<!-- Sent from Reviewable.io -->

tamird commented:
You could imagine wanting a shorter one if you had a deadline, or you could
want to plumb in a context for reasons other than lifetime, like logging or
tracing.
On Mar 24, 2016 17:17, "Tobias Schottdorf" <notifications@github.com> wrote:

> Review status: all files reviewed at latest revision, 4 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *client/db.go, line 412 [r3]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeMUkQoQe84giKsrhF:-KDeUEwmKVjPakSMN4HX:62476964>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/client/db.go#L412>):*
> Why? The only lifetime of a context that makes sense with the current
> behavior is one for the lifetime of the transaction, and that's what we
> have here.
> ------------------------------
>
> *Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/5447>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5447#issuecomment-201024296>
>

paperstreet commented:
Added a TODO and some docs about context semantics

---

Review status: 7 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[client/db.go, line 412 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeMUkQoQe84giKsrhF:-KDeSSFHB7fi26pa_Moe:2050576819) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/client/db.go#L412)):*
This breaks some tests. I added a TODO and I'll do it in a followup.

---

*[kv/txn_coord_sender_test.go, line 356 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeLAWireRrgkBpBaEL:-KDeZ7AW0MWviagrB_2G:-815485757) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/kv/txn_coord_sender_test.go#L356)):*
Done. Merged it with the other loop

---

*[kv/txn_coord_sender_test.go, line 571 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeNSSMO1V1k3sh5cTy:-KDeZ29jKpx010sBWbG2:156453455) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/kv/txn_coord_sender_test.go#L571)):*
Added a TODO and made the timeout shorter

---

*[kv/txn_coord_sender_test.go, line 1270 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDeLVL4SskJk70HiUtn:-KDeUpQjAufpwScx0NRZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/35042d8be0a98bf2b6cb73fbc0aac82bb0c3ce3d/kv/txn_coord_sender_test.go#L1270)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-:-KDeZueUteeoEJuuPWOW:316822257)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[kv/txn_coord_sender_test.go, line 559 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDfEHRUne3lyLyZvTyI:-KDfEHRV_S4KRrBbze5Q:-1512380947) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abc99a086a39157daf655a5dab77639f3b8b9ee/kv/txn_coord_sender_test.go#L559)):*
what's the point of `stilAlive`? anywhere you set it to `false` you can just early return `nil`.

---

*[kv/txn_coord_sender_test.go, line 564 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDfEYrd3ZyGC-CZkfly:-KDfEYrd3ZyGC-CZkflz:175771959) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abc99a086a39157daf655a5dab77639f3b8b9ee/kv/txn_coord_sender_test.go#L564)):*
this should be t.Fatal i think - the error is not likely to be transient

---

*[kv/txn_coord_sender_test.go, line 574 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDfEhCPAy4nY6V-cNC3:-KDfEhCPAy4nY6V-cNC4:16092222) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abc99a086a39157daf655a5dab77639f3b8b9ee/kv/txn_coord_sender_test.go#L574)):*
s/%s/%v/ because `err` might be nil

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
Review status: 9 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[kv/txn_coord_sender_test.go, line 559 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDfEHRUne3lyLyZvTyI:-KDfG9kWz3MJQn5KP7Hd:-518967476) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abc99a086a39157daf655a5dab77639f3b8b9ee/kv/txn_coord_sender_test.go#L559)):*
Yeah, that's much better.

---

*[kv/txn_coord_sender_test.go, line 564 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDfEYrd3ZyGC-CZkfly:-KDfG9MRBji7jL5mzdwE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abc99a086a39157daf655a5dab77639f3b8b9ee/kv/txn_coord_sender_test.go#L564)):*
Done.

---

*[kv/txn_coord_sender_test.go, line 574 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447#-KDfEhCPAy4nY6V-cNC3:-KDfGCuyHas05ZwGuKzb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abc99a086a39157daf655a5dab77639f3b8b9ee/kv/txn_coord_sender_test.go#L574)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5447)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
:thumbsup:!

On Thu, Mar 24, 2016 at 10:27 PM Dan Harrison <notifications@github.com>
wrote:

> Merged #5447 <https://github.com/cockroachdb/cockroach/pull/5447>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5447#event-602595263>
>
